src/sbf.cpp

```diff
@@ -0,0 +1,579 @@
+			for (int i = 0; i < ret; i++) {
+				handled |= parseChar(buf[i]);
+				// SBF_DEBUG("parsed %d: 0x%x", i, buf[i]);
+			}
+		}
+
+		if (handled > 0) {
+			return handled;
+		}
+
+		// abort after 2s if no useful packets received
+		if (time_started + 2000 * 1000 < gps_absolute_time()) {
+			SBF_DEBUG("timed out, returning");
+			return -1;
+		}
+	}
+}
+
+int    // 0 = decoding, 1 = message handled, 2 = sat info message handled
+GPSDriverSBF::parseChar(const uint8_t b)
+{
+	int ret = 0;
+
+	switch (_decode_state) {
+
+	// Expecting Sync1
+	case SBF_DECODE_SYNC1:
+		if (b == SBF_SYNC1) { // Sync1 found --> expecting Sync2
+			SBF_TRACE_PARSER("A");
+			ret = payloadRxAdd(b); // add a payload byte
+			_decode_state = SBF_DECODE_SYNC2;
+
+		} else if (b == RTCM3_PREAMBLE && _rtcm_parsing) {
+			SBF_TRACE_PARSER("RTCM");
+			_decode_state = SBF_DECODE_RTCM3;
+			_rtcm_parsing->addByte(b);
+		}
+
+		break;
+
+	// Expecting Sync2
+	case SBF_DECODE_SYNC2:
+		if (b == SBF_SYNC2) { // Sync2 found --> expecting CRC
+			SBF_TRACE_PARSER("B");
+			ret = payloadRxAdd(b); // add a payload byte
+			_decode_state = SBF_DECODE_PAYLOAD;
+
+		} else { // Sync1 not followed by Sync2: reset parser
+			decodeInit();
+		}
+
+		break;
+
+	// Expecting payload
+	case SBF_DECODE_PAYLOAD:
+		SBF_TRACE_PARSER(".");
+
+		ret = payloadRxAdd(b); // add a payload byte
+
+		if (ret < 0) {
+			// payload not handled, discard message
+			ret = 0;
+			decodeInit();
+
+		} else if (ret > 0) {
+			ret = payloadRxDone(); // finish payload processing
+			decodeInit();
+
+		} else {
+			// expecting more payload, stay in state SBF_DECODE_PAYLOAD
+			ret = 0;
+
+		}
+
+		break;
+
+	case SBF_DECODE_RTCM3:
+		if (_rtcm_parsing->addByte(b)) {
+			SBF_DEBUG("got RTCM message with length %i", (int)_rtcm_parsing->messageLength());
+			gotRTCMMessage(_rtcm_parsing->message(), _rtcm_parsing->messageLength());
+			decodeInit();
+		}
+
+		break;
+	}
+
+	return ret;
+}
+
+/**
+ * Add payload rx byte
+ */
+int    // -1 = error, 0 = ok, 1 = payload completed
+GPSDriverSBF::payloadRxAdd(const uint8_t b)
+{
+	int ret = 0;
+	uint8_t *p_buf = reinterpret_cast<uint8_t *>(&_buf);
+
+	p_buf[_rx_payload_index++] = b;
+
+	if ((_rx_payload_index > 7 && _rx_payload_index >= _buf.length) || _rx_payload_index >= sizeof(_buf)) {
+		ret = 1; // payload received completely
+	}
+
+	return ret;
+}
+
+/**
+ * Calculate buffer CRC16
+ */
+uint16_t
+crc16(const uint8_t *data_p, uint32_t length)
+{
+	uint8_t x;
+	uint16_t crc = 0;
+
+	while (length--) {
+		x = crc >> 8 ^ *data_p++;
+		x ^= x >> 4;
+		crc = static_cast<uint16_t>((crc << 8) ^ (x << 12) ^ (x << 5) ^ x);
+	}
+
+	return crc;
+}
+
+/**
+ * Finish payload rx
+ */
+int    // 0 = no message handled, 1 = message handled, 2 = sat info message handled
+GPSDriverSBF::payloadRxDone()
+{
+	int ret = 0;
+	struct tm timeinfo;
+	time_t epoch;
+	uint8_t *buf_ptr;
+
+	if (_buf.crc16 != crc16(reinterpret_cast<uint8_t *>(&_buf) + 4, _buf.length - 4)) {
```

Comment:
We should check for `_buf.length >= 4` to avoid invalid memory access on corrupt input.

Reply:
Implemented.